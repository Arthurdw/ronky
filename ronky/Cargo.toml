[package]
name = "ronky"
description = "A simple way to export Rust definitions to Arri types"
readme = "README.md"

version.workspace = true
authors.workspace = true
categories.workspace = true
edition.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

[lints]
workspace = true

[dependencies]
arri_repr = { workspace = true, features = [] }
ronky_derive = { workspace = true, optional = true }

[dev-dependencies]
ronky_derive.workspace = true

[features]
default = []
derive = ["ronky_derive"]
chrono = ["arri_repr/chrono"]
time = ["arri_repr/time"]
uuid = ["arri_repr/uuid"]
bigdecimal = ["arri_repr/bigdecimal"]
num-bigint = ["arri_repr/num-bigint"]
num-bigfloat = ["arri_repr/num-bigfloat"]
rust_decimal = ["arri_repr/rust_decimal"]
url = ["arri_repr/url"]
decimal = ["arri_repr/decimal"]
bytes = ["arri_repr/bytes"]
dashmap = ["arri_repr/dashmap"]
indexmap = ["arri_repr/indexmap"]
smallvec = ["arri_repr/smallvec"]
