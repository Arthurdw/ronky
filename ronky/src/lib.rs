//! # üê± RONKY üê±
//!
//! [![Crates.io Version](https://img.shields.io/crates/v/ronky)](https://crates.io/crates/ronky)
//! [![Purrs Per Minute](https://img.shields.io/badge/purrs-over%209000-orange)](https://github.com/modiimedia/arri)
//! [![Cat Approved](https://img.shields.io/badge/cat-approved-brightgreen)](https://github.com/modiimedia/arri)
//! [![Rustacean Friendly](https://img.shields.io/badge/rustacean-friendly-blue)](https://github.com/modiimedia/arri)
//!
//! _"Converting Rust types shouldn't be this purr-fect, but here we are..."_
//!
//! ## üò∫ What in the Whiskers is Ronky?
//!
//! Imagine if your Rust types could speak other languages without learning a single foreign word.
//! That's Ronky ‚Äì your code's personal polyglot translator that speaks fluent
//! [Arri](https://github.com/modiimedia/arri), turning your carefully crafted Rust types into
//! schemas that even JavaScript developers can understand.
//!
//! Born from the frustration of manual schema creation (and named after a particularly vocal cat),
//! Ronky does the tedious work so you can focus on the important stuff ‚Äì like deciding whether your
//! next variable should be called `data` or `info` (we both know you'll pick `data`).
//!
//! > üöß **Paws at Work**: Like a cat that's not quite finished knocking everything off your desk,
//! > Ronky is still under construction. Object serialization and deserialization are coming soon,
//! > probably right after this catnap. üöß
//!
//! ## ‚ú® Features That Make You Go "Meow!"
//!
//! Ronky doesn't just toss your types over the fence to Arri-land. It crafts them with the same
//! attention to detail that a cat gives to knocking your most precious possessions off shelves:
//!
//! ```text
//! ‚îå‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îê
//! ‚îÇ                          RONKY'S REPERTOIRE                           ‚îÇ
//! ‚îú‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î§
//! ‚îÇ üß¨ Type Wizardry          ‚îÇ - Transforms primitives with grace        ‚îÇ
//! ‚îÇ                           ‚îÇ - Handles generic types without whining   ‚îÇ
//! ‚îÇ                           ‚îÇ - Makes associated types feel welcome     ‚îÇ
//! ‚îú‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î§
//! ‚îÇ üß© Collection Conjuring   ‚îÇ - Vectors become elegant "elements"       ‚îÇ
//! ‚îÇ                           ‚îÇ - Maps manifest as "values" schemas       ‚îÇ
//! ‚îÇ                           ‚îÇ - Optional types know when to disappear   ‚îÇ
//! ‚îú‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î§
//! ‚îÇ üõ°Ô∏è Guardian Features      ‚îÇ - Strict mode keeps schemas pristine      ‚îÇ
//! ‚îÇ                           ‚îÇ - Discriminators tag unions properly      ‚îÇ
//! ‚îÇ                           ‚îÇ - Circular refs handled without dizziness ‚îÇ
//! ‚îú‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î§
//! ‚îÇ üîÑ Transformation Magic   ‚îÇ - Case transformations (snake ‚Üí UPPER)    ‚îÇ
//! ‚îÇ                           ‚îÇ - Field renaming for multilingual joy     ‚îÇ
//! ‚îÇ                           ‚îÇ - Nullable marking for optional presence  ‚îÇ
//! ‚îú‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î§
//! ‚îÇ üìù Documentation Delight  ‚îÇ - Comments become documentation           ‚îÇ
//! ‚îÇ                           ‚îÇ - Deprecation warnings that don't nag     ‚îÇ
//! ‚îÇ                           ‚îÇ - Metadata that brings joy to readers     ‚îÇ
//! ‚îî‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îò
//! ```
//!
//! With Ronky, your type schema generation is both:
//! 1. **Compile-time verified** - Errors at compile time, not at 3 AM when you're deploying
//! 2. **Automatically generated** - Because life's too short to manually update schemas
//!
//! > üí° **Pro Tip**: Ronky's powers grow with your documentation. The more doc comments you add,
//! > the more magnificent your schemas become. It's like feeding treats to a cat ‚Äì the rewards
//! > are well worth it.
//!
//! ## ü§ù The Cool Cats Club (Compatible Crates)
//!
//! Ronky has an extensive social network. Think of these crates as the neighborhood cats that
//! regularly visit your backyard ‚Äì they're all welcome and get special treatment:
//!
//! ```text
//! TEMPORAL FRIENDS         üï∞Ô∏è  chrono, time
//! IDENTITY SPECIALISTS     ü™™  uuid
//! BIG NUMBER EXPERTS       üî¢  bigdecimal, num-bigint, num-bigfloat
//! PRECISION MASTERS        üí∞  rust_decimal, decimal
//! WEB-SAVVY NAVIGATORS     üåê  url
//! DATA-HANDLING WIZARDS    üìä  bytes
//! CONCURRENT COMPANIONS    üßµ  dashmap
//! ORDERLY ORGANIZERS       üìã  indexmap
//! OPTIMIZED PERFORMERS     ‚ö°  smallvec
//! ```
//!
//! Each of these crates gets the VIP (Very Important Purring) treatment from Ronky. Their types
//! are handled with the care and respect they deserve.
//!
//! > üêà **Missing your favorite companion?** Open an issue to suggest a new addition to Ronky's
//! > compatible crates collection. The more the merrier!
//!
//! ## üë®‚Äçüíª The Story Behind Ronky
//!
//! Let me share the real origin story ‚Äì Ronky wasn't born from late-night caffeine-fueled frustration
//! (though there's been plenty of that in my coding career). It all started when a friend dropped a
//! Discord message with a link to the [Arri project](https://github.com/modiimedia/arri). One look at
//! what Arri was doing with type conversions and I was hooked.
//!
//! As a passionate Rustacean, I immediately thought: "This is brilliant! But it needs Rust support."
//! And thus Ronky was born ‚Äì creating a bridge between the elegant type system of Rust and the universal
//! schema language of Arri.
//!
//! I named it after my cat not just because he was adorable (though he absolutely was), but because
//! he embodied what good libraries should be ‚Äì helpful, reliable, and occasionally surprising you
//! with something delightful. Like Ronky the cat finding innovative ways to get treats without
//! moving, Ronky the library finds ways to update your API schemas without you lifting a finger.
//!
//! My hope is that this library saves you time, prevents those pesky "the API contract changed but
//! the docs didn't" bugs, and maybe ‚Äì just maybe ‚Äì frees up enough of your day to spend time with
//! your own four-legged friend (or fish, or rubber duck - whatever brings you joy).
//!
//! ## üìö The Illustrated Guide to Ronky
//!
//! ### üîÑ The Basic Transformation
//!
//! ```rust
//! use ronky::{Exportable, Exported, SCHEMA_VERSION};
//! use serde_json::{Value, from_str, to_string_pretty};
//!
//! // Just add water (and a derive macro)
//! #[derive(Exported)]
//! #[arri(transform = "uppercase")] // LOUD NOISES
//! enum Result<T: Exportable, E: Exportable> {
//!     /// When things go right (rarely, if you're me)
//!     Ok(T),
//!     /// When things go wrong (my default state)
//!     Err(E),
//! }
//!
//!
//! // Announce our intentions to the world
//! println!("üß™ Creating an Arri {} schema and hoping for the best...", SCHEMA_VERSION);
//!
//! // The cat-alchemy happens here
//! let schema_json = Result::<String, ()>::export()
//!     .serialize()
//!     .expect("this to work (please, I have deadlines)");
//!
//! // Humans like pretty things
//! let pretty_json = to_string_pretty(&from_str::<Value>(&schema_json).unwrap()).unwrap();
//!
//! // Admire our handiwork
//! println!("{}", pretty_json);
//!
//! // Now go make a cup of tea, you've earned it
//! ```
//!
//! ### üß© The Advanced Cat-egory: Building Complex Types
//!
//! ```rust
//! use ronky::{Exportable, Exported, SCHEMA_VERSION};
//!
//! /// Metadata about things (and sometimes other things)
//! #[derive(Exported)]
//! struct About<T: Exportable> {
//!     /// What we called it before marketing got involved
//!     #[deprecated(since = "1.0.0", note = "Use `firstName` and `lastName` instead")]
//!     name: String,
//!     
//!     /// The name that appears on your coffee cup at Starbucks
//!     first_name: String,
//!     
//!     /// The name your parents use when you're in trouble
//!     last_name: Option<String>,
//!     
//!     /// The number that makes you sigh at birthday parties
//!     age: u32,
//!     
//!     /// The subject of our obsession
//!     of: T,
//! }
//!
//! /// A creature that creates Rust crates, ironically
//! #[derive(Exported)]
//! #[arri(strict)] // No surprises allowed! Like a cat with a cucumber
//! struct Human {
//!     /// Fellow code-monkeys who review your PRs
//!     friends: Vec<Human>, // Recursive types? No problem!
//!     
//!     /// The real owners of your home
//!     pets: Vec<About<Pet>>,
//! }
//!
//! /// Fashion choices for the discerning feline
//! #[derive(Exported)]
//! #[arri(transform = ["snake_case", "uppercase"])] // MULTI_STYLE_TRANSFORMATION
//! enum CatColor {
//!     /// Like my coffee and my humor
//!     Black,
//!     
//!     /// Like my documentation standards and error handling
//!     White,
//!     
//!     /// Like my moral compass when it comes to optimization
//!     Gray,
//!     
//!     /// Like my commit history after a weekend hackathon
//!     MixedGrayWhite,
//! }
//!
//! /// Entities that interrupt your Zoom calls at the worst possible moment
//! #[derive(Exported)]
//! #[arri(transform = "uppercase", discriminator = "species")]
//! enum Pet {
//!     Dog {
//!         /// The word you'll repeat 37 times at the dog park
//!         name: String,
//!         
//!         /// What you'll forget when the vet asks
//!         #[arri(nullable)]
//!         breed: Option<String>,
//!     },
//!     
//!     #[arri(rename = "cat")] // All hail the cat overlords!
//!     Lion {
//!         /// A suggestion they might consider responding to someday
//!         name: String,
//!         
//!         /// Their royal garment
//!         #[arri(nullable)]
//!         color: Option<CatColor>,
//!     },
//! }
//! ```
//!
//! > üî• **Hot Tip**: These examples aren't just decorative ‚Äì they're functional!
//! > Copy, paste, and experience the magic of Ronky firsthand. Your future self
//! > will thank you when your API documentation is automatically up-to-date.
//!
//! ## üêà The Ronky Memorial Section
//!
//! ```text
//!      /\_/\  
//!     ( o.o )
//!      > ^ <
//!     /  O  \  "Meow meow, transform types meow."
//!                                     - Ronky (2010-2024)
//! ```
//!
//! This library immortalizes a magnificent cat named Ronky, whose thunderous purrs
//! (or "ronks" in Dutch) could wake the neighbors. For 14 remarkable years, this
//! whiskered genius supervised everything that happened in the house.
//!
//! Despite battling acromegaly, Ronky maintained a proud dignity and an uncanny ability
//! to walk across keyboards. His legacy continues in this library!
//!
//! Ronky taught me important programming principles:
//!
//! - **Persistence**: If at first you don't succeed, meow louder until someone fixes it for you
//! - **Efficiency**: Why do something yourself when you can delegate?
//! - **Work-Life Balance**: No bug is so important that it can't wait until after a nap
//! - **Code Reviews**: Sometimes the best feedback is just silently judging from a distance
//!
//! This library aims to embody these principles by automating the tedious parts of
//! API development so you can focus on the parts that matter ‚Äì like figuring out why
//! your application works in development but not production (it's always CORS).
//!
//! ## üìã Quick Reference
//!
//! ### The Basics
//!
//! 1. Add `ronky` to your `Cargo.toml`:
//!    ```toml
//!    [dependencies]
//!    ronky = "1.0.0"  # Check crates.io for the latest version
//!    ```
//!
//! 2. Import the essentials:
//!    ```rust
//!    use ronky::{Exported, SCHEMA_VERSION};
//!    ```
//!
//! 3. Decorate your types:
//!    ```rust
//!    #[derive(Exported)]
//!    struct MyType { /* fields */ }
//!    ```
//!
//! 4. Export and serialize:
//!    ```rust
//!    let schema = MyType::export().serialize().unwrap();
//!    ```
//!
//! 5. Profit! (This step is not automated by Ronky, sorry)
//!
//! ### Attribute Options
//!
//! - `#[arri(strict)]` - No extra properties allowed
//! - `#[arri(transform = "snake_case")]` - Transform enum variant names
//! - `#[arri(discriminator = "type")]` - Set discriminator field name
//! - `#[arri(rename = "newName")]` - Rename a field or variant
//! - `#[arri(nullable)]` - Mark a field as nullable
//!
//! ## üåü Final Thought
//!
//! Remember: Type conversion should be like a cat's nap ‚Äì automatic, elegant, and requiring
//! no effort on your part. Let Ronky handle the tedious work while you focus on building
//! something amazing.
//!
//! Now go pet your cat (or dog, or rubber duck) ‚Äì they've been waiting patiently while you
//! read this documentation. ‚ù§Ô∏è

#[cfg(feature = "derive")]
extern crate ronky_derive;

#[cfg(feature = "derive")]
pub use ronky_derive::Exported;

extern crate arri_repr;
pub use arri_repr::*;

pub static SCHEMA_VERSION: &str = "v0.0.8";

// TODO: implement conversion from ATD to Rust types
// | ATD Type | Rust Type |
// |---|---|
// | string | String |
// | boolean | bool |
// | timestamp | DateTime |
// | float32 | f32 |
// | float64 | f64 |
// | int8 | i8 |
// | uint8 | u8 |
// | int16 | i16 |
// | uint16 | u16 |
// | int32 | i32 |
// | uint32 | u32 |
// | int64 | i64 |
// | uint64 | u64 |
