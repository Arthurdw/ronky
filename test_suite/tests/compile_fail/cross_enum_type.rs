use ronky::Exported;

#[derive(Exported)]
enum Mixed {
    Normal,
    Tagged { name: String },
}

fn main() {}
