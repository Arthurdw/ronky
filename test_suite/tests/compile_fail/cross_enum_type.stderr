error: Arri requires that enums can only be all enum or all tagged union variants. This variant violates that rule.
 --> tests/compile_fail/cross_enum_type.rs:6:5
  |
6 |     Tagged { name: String },
  |     ^^^^^^
